### DISCUSSION TITLE
Windscribe Static IP with Port Forward

### DISCUSSION THREAD

**Discussion by ShieldsPC** [qdm12/gluetun#1499] [Q&A] [answered]
Basically, I want to use a Gluetun container as my gateway for some other containers. However, I want to do it with my static IP from Windscribe. I also want to forward port 12345. So I want a container to be accessible at port 12345 through my Windscribe static IP. I can create a manual device on Windscribe and define a port, but it gives me some OpenVPN information that I don't see a space for in the Gluetun compose file. "Connect IP" is not asked for when creating an OpenVPN Gluetun container. I also don't see a way to tell it what ports I want to forward. Any advice?

**Comment by bnhf** [comment]
@ShieldsPC 

If you have a .ovpn file that works using the OpenVPN GUI or other client, you could always try the custom provider option in Gluetun. I stepped through this recently (#1470) with someone wanting to connect to their own OpenVPN Access Server. There's a bunch of back-and-forth you can ignore due to an inaccurate .ovpn file being posted initially (comment "#" marks removed for some reason), but the ultimate answer might be workable for you.

**Reply by ShieldsPC** [reply]
This might be what I'm looking for! Is there a flag in the Docker compose to tell it where the .ovpn file is? Sorry if this is a dumb question. I'm brand new to Docker.

**Comment by bnhf** [comment]
@ShieldsPC 

If you look at Docker compose below, you'll see that the name of the .ovpn file is established with OPENVPN_CUSTOM_CONFIG, and then that file on the host computer is bound to the /gluetun/yourfilename.ovpn:ro file in the container. Use of the VPN provider "custom" tells gluetun that you're going to be using your own config file:

```yml
version: '3.7'
services:
  gluetun:
    image: qmcgaw/gluetun:latest
    container_name: gluetun
    cap_add:
      - NET_ADMIN
    environment:
      - VPN_SERVICE_PROVIDER=custom
      - VPN_TYPE=openvpn
      - OPENVPN_CUSTOM_CONFIG=/gluetun/xxx.ovpn # Replace with the actual name of your .ovpn file
      - OPENVPN_USER=openvpn # OpenVPN AS username here (openvpn is the default user)
      - OPENVPN_PASSWORD=xxx # Password associated with the username
      - TZ=US/Mountain # Timezone in standard Linux format
    volumes:
      - /data/openvpnas/xxx.ovpn:/gluetun/xxx.ovpn:ro # Bind the directory/filename to /gluetun/xxx.ovpn here with ro option
      - /data/openvpnas:/gluetun # /data/openvpnas will be used for servers.json
  ```

**Comment by ShieldsPC** [comment]
@bnhf
Progress! Using your Docker compose, I get a stack that has a Gluetun container that reports the correct IP address. However, when I point my other containers to it using the network tab in Portainer and then get the IP of those containers, it shows my local IP, not my VPN IP. Any thoughts?

**Comment by bnhf** [comment, marked as answer]
@ShieldsPC 

If you're using separate stacks for your other containers, you should structure them to only use Gluetun to get the Internet. Here's an example of how it would be done with a Firefox container (something I like to use for testing). It may be possible to do it through the Portainer network tab, but this way there are no extra steps:

```yml
version: '3.7'
services:
  firefox:
    image: lscr.io/linuxserver/firefox:latest
    container_name: firefox
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=US/Mountain
    volumes:
      - /data/firefox:/config
    shm_size: '1gb'
    network_mode: 'container:gluetun'
```

All of your Gluetun dependent containers can also be put in the same stack, which I prefer -- but either way works. In the same stack, the networking section for each service (container) would look like this:

```yml
    network_mode: 'service:gluetun'
    depends_on:
      - gluetun
```

And any ports you need to open, for any container in the stack, get listed in the Gluetun section.

**Reply by ShieldsPC** [reply]
Thank you so much! I'll do my homework on Docker networking.

**Comment by jordantrizz** [comment]
I wanted to post as I ran into the same issue. The following section helped me https://github.com/qdm12/gluetun/wiki/Windscribe#vpn-server-port-forwarding setting FIREWALL_VPN_INPUT_PORTS to the port that I set up to forward. It started working.

**Reply by bnhf** [reply]
@jordantrizz 

> I wanted to post as I ran into the same issue.

So you are also using a static IP from Windscribe (extra cost option) as described here: https://windscribe.com/staticips?

**Reply by jordantrizz** [reply]
Yes, and got it working.

**Reply by jordantrizz** [reply]
```yml
services:
  gluetun:
    image: qmcgaw/gluetun
    container_name: gluetun
    cap_add:
      - NET_ADMIN
    environment:
      - VPN_SERVICE_PROVIDER=custom # Using custom .ovpn file
      - VPN_TYPE=openvpn
      - OPENVPN_USER=14123123
      - OPENVPN_PASSWORD=14123123
      - OPENVPN_CUSTOM_CONFIG=/gluetun/windscribe.ovpn # Custom ovpn file
      - FIREWALL_VPN_INPUT_PORTS=42001,42001,6881 # Open up ports on openvpn
      # - HTTP_CONTROL_SERVER_ADDRESS=:8082 # Was 8080 but had a conflict.
    volumes:
      - $PWD/windscribe.ovpn:/gluetun/windscribe.ovpn:ro
```

**Reply by bnhf** [reply]
@jordantrizz 

Thanks for posting your docker-compose, that helps to clarify what you were getting at with your original comment.

So just to summarize, you're confirming the need to use a custom .ovpn for those that are signed-up for a static IP through Windscribe. In addition though, you're making an excellent point regarding port forwarding (which would be quite common for those signed-up for the static IP service), and the need to use the FIREWALL_VPN_INPUT_PORTS environment variable to let specific port traffic through the Gluetun firewall. Those ports, of course, would be the same ones you opened through the Windscribe account management page.

When you have a moment, would you mind posting the .ovpn you're using (with sensitive info redacted)? @qdm12 may want to add the environment variables required to support a Windscribe static IP natively in a future update -- so it'd be great to confirm which OpenVPN directives are being used.

**Reply by jordantrizz** [reply]
Sorry, I should have explained it better. I found this thread due to my own issues.



> So just to summarize, you're confirming the need to use a custom .ovpn for those that are signed-up for a static IP through Windscribe. In addition though, you're making an excellent point regarding port forwarding (which would be quite common for those signed-up for the static IP service), and the need to use the FIREWALL_VPN_INPUT_PORTS environment variable to let specific port traffic through the Gluetun firewall. Those ports, of course, would be the same ones you opened through the Windscribe account management page.

Correct.



> When you have a moment, would you mind posting the .ovpn you're using (with sensitive info redacted)? @qdm12 may want to add the environment variables required to support a Windscribe static IP natively in a future update -- so it'd be great to confirm which OpenVPN directives are being used.

```
client
dev tun
proto udp
remote 37.120.141.131 443
verify-x509-name ams-120.windscribe.com name

nobind
auth-user-pass

resolv-retry infinite

cipher AES-256-GCM
ncp-ciphers AES-256-GCM:AES-256-CBC:AES-128-GCM
auth SHA512

verb 2
mute-replay-warnings
remote-cert-tls server
persist-key
persist-tun

key-direction 1
<ca>
-----BEGIN CERTIFICATE-----
CERT CONTENT
-----END CERTIFICATE-----
</ca>
<tls-auth>
-----BEGIN OpenVPN Static key V1-----
KEY CONTENT
-----END OpenVPN Static key V1-----
</tls-auth>
```

**Reply by MrInterBugs** [reply]
So I was using this set up for weeks and suddenly its stopped working, any idea's why? is everyone's else's windscribe still working?

The logs look normal:
```
2023-04-25T21:19:29+01:00 INFO [routing] default route found: interface eth0, gateway 172.19.0.1 and assigned IP 172.19.0.2
2023-04-25T21:19:29+01:00 INFO [routing] adding route for 0.0.0.0/0
2023-04-25T21:19:29+01:00 INFO [firewall] setting allowed subnets...
2023-04-25T21:19:29+01:00 INFO [routing] default route found: interface eth0, gateway 172.19.0.1 and assigned IP 172.19.0.2
2023-04-25T21:19:29+01:00 INFO [routing] adding route for 172.20.0.0/16
2023-04-25T21:19:29+01:00 INFO [routing] adding route for 192.168.1.0/24
2023-04-25T21:19:29+01:00 INFO [dns over tls] using plaintext DNS at address 1.1.1.1
2023-04-25T21:19:29+01:00 INFO [http server] http server listening on [::]:8000
2023-04-25T21:19:29+01:00 INFO [healthcheck] listening on 127.0.0.1:9999
2023-04-25T21:19:29+01:00 INFO [firewall] allowing VPN connection...
2023-04-25T21:19:29+01:00 INFO [openvpn] 2023-04-25 21:19:29 Note: Treating option '--ncp-ciphers' as  '--data-ciphers' (renamed in OpenVPN 2.5).
2023-04-25T21:19:29+01:00 INFO [openvpn] OpenVPN 2.5.8 aarch64-alpine-linux-musl [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on Nov  2 2022
2023-04-25T21:19:29+01:00 INFO [openvpn] library versions: OpenSSL 3.0.8 7 Feb 2023, LZO 2.10
2023-04-25T21:19:29+01:00 INFO [openvpn] TCP/UDP: Preserving recently used remote address: [AF_INET]xxxxxxxxx:443
2023-04-25T21:19:29+01:00 INFO [openvpn] UDP link local: (not bound)
2023-04-25T21:19:29+01:00 INFO [openvpn] UDP link remote: [AF_INET]xxxxxxxxxxx:443
2023-04-25T21:19:29+01:00 INFO [openvpn] [xxxxxxx.windscribe.com] Peer Connection Initiated with [AF_INET]xxxxxxxxxx:443
2023-04-25T21:19:31+01:00 INFO [openvpn] TUN/TAP device tun0 opened
2023-04-25T21:19:31+01:00 INFO [openvpn] /sbin/ip link set dev tun0 up mtu 1500
2023-04-25T21:19:31+01:00 INFO [openvpn] /sbin/ip link set dev tun0 up
2023-04-25T21:19:31+01:00 INFO [openvpn] /sbin/ip addr add dev tun0 10.117.234.4/23
2023-04-25T21:19:31+01:00 INFO [openvpn] UID set to nonrootuser
2023-04-25T21:19:31+01:00 INFO [openvpn] Initialization Sequence Completed
2023-04-25T21:19:31+01:00 INFO [firewall] setting allowed input port xxxxx through interface tun0...
2023-04-25T21:19:31+01:00 INFO [dns over tls] downloading DNS over TLS cryptographic files
2023-04-25T21:19:31+01:00 INFO [healthcheck] healthy!
2023-04-25T21:19:34+01:00 INFO [dns over tls] downloading hostnames and IP block lists
2023-04-25T21:19:49+01:00 WARN [dns over tls] context deadline exceeded (Client.Timeout or context cancellation while reading body)
2023-04-25T21:19:49+01:00 INFO [dns over tls] init module 0: validator
2023-04-25T21:19:49+01:00 INFO [dns over tls] init module 1: iterator
2023-04-25T21:19:49+01:00 INFO [dns over tls] start of service (unbound 1.17.1).
2023-04-25T21:19:49+01:00 INFO [dns over tls] generate keytag query _ta-4a5c-4f66. NULL IN
2023-04-25T21:19:49+01:00 INFO [dns over tls] generate keytag query _ta-4a5c-4f66. NULL IN
2023-04-25T21:19:50+01:00 INFO [dns over tls] ready
2023-04-25T21:19:50+01:00 INFO [ip getter] Public IP address is xxxxxxxxxxxxxxxxxxxxx
2023-04-25T21:19:50+01:00 INFO [vpn] You are running 1 commit behind the most recent latest
```

![image](https://user-images.githubusercontent.com/21267797/234396445-be55f4a3-1445-47ca-81e7-0156c16df478.png)

Finally:
```
 gluetun:
    image: qmcgaw/gluetun:latest
    restart: always
    container_name: gluetun
    cap_add:
      - NET_ADMIN
    devices:
      - /dev/net/tun:/dev/net/tun
    ports:
      - 8080:8080/tcp
    volumes:
      - ./ovpn.conf:/gluetun/custom.conf:ro
    environment:
      - VPN_SERVICE_PROVIDER=custom
      - VPN_TYPE=openvpn
      - OPENVPN_CUSTOM_CONFIG=/gluetun/custom.conf
      - OPENVPN_USER=xxxxxxx
      - OPENVPN_PASSWORD=xxxxxxxx
      - FIREWALL_OUTBOUND_SUBNETS=192.168.1.0/24
      - FIREWALL_VPN_INPUT_PORTS=xxxxx
      - TZ=Europe/London
```

**Reply by MomentusCritisae** [reply]
I just wanted to add that I faced the same problem recently and this docker-compose worked for me. I also have a static IP with port forwarding.

> ```yaml
> services:
>   gluetun:
>     image: qmcgaw/gluetun
>     container_name: gluetun
>     cap_add:
>       - NET_ADMIN
>     environment:
>       - VPN_SERVICE_PROVIDER=custom # Using custom .ovpn file
>       - VPN_TYPE=openvpn
>       - OPENVPN_USER=14123123
>       - OPENVPN_PASSWORD=14123123
>       - OPENVPN_CUSTOM_CONFIG=/gluetun/windscribe.ovpn # Custom ovpn file
>       - FIREWALL_VPN_INPUT_PORTS=42001,42001,6881 # Open up ports on openvpn
>       # - HTTP_CONTROL_SERVER_ADDRESS=:8082 # Was 8080 but had a conflict.
>     volumes:
>       - $PWD/windscribe.ovpn:/gluetun/windscribe.ovpn:ro
> ```

**Reply by EricSeastrand** [reply]
@ShieldsPC Unless you have any objection, I think [the comment(s) above](https://github.com/qdm12/gluetun/discussions/1499#discussioncomment-5605757) should be marked as the "Answer" for this question. Best I can tell, this is the solution to the problem you faced with static IP/port forward/etc..

Can you update the answer? It'll help the next lucky Googler to find this answer faster :)

**Comment by dnham314** [comment]
Has anyone been able to get it to work with Wireguard?

**Reply by EricSeastrand** [reply]
Yes- see above https://github.com/qdm12/gluetun/discussions/1499#discussioncomment-9866377

Unless you mean: did anybody figure out how to make it auto-regen the port after windscribe expires your port reservation, update the container configs to the new port, restart, etc… Sadly no, that’s still manual.
If anyone figures that out… hmu 😁

**Reply by dnham314** [reply]
I'm interested in static ip port forwarding with a custom Wireguard config.
